core/platform/src/raw_syscalls.rs

```diff
@@ -3,98 +3,94 @@
+// 2.0 syscall TRD. Registers r0-r4 correspond to ARM registers r0-r4 and RISC-V
+// registers a0-a4.
+//
+// Theoretically, RawSyscalls could consist of a single raw system call. To
+// start, something like this should work:
+//
+//   unsafe fn syscall<const CLASS: usize>([usize; 4]) -> [usize; 4];
+//
+// However, this will not work with Miri's -Zmiri-track-raw-pointers flag, as it
+// causes pointers passed to the kernel via the Allow system calls to be
+// untagged. In order to work with -Zmiri-track-raw-pointers, we need to pass
+// pointers for the register values. Rust's closest analogue to C's void pointer
+// is *mut () or *const (); we use *mut () because it is shorter:
+//
+//   unsafe fn syscall<const CLASS: usize>([*mut (); 4]) -> [*mut (); 4];
+//
+// Using a single system call has a major inefficiency. The single raw system
+// call would need to clobber every register that any system call can clobber.
+// Yield has a far longer clobber list than most system calls, so this would be
+// inefficient for the majority of system calls. As a result, we can split yield
+// out into its own function, giving the following API:
+//
+//   unsafe fn yield([*mut (); 4]) -> [*mut (); 4];
+//   unsafe fn syscall<const CLASS: usize>([*mut (); 4]) -> [*mut (); 4];
+//
+// There is one significant inefficiency remaining. Many system calls, such as
+// memop's "get RAM start address" operation, do not need to set all four
+// arguments. The compiler cannot optimize away this inefficiency, so to remove
+// it we need to split the system calls up based on the number of arguments they
+// take:
+//
+//   unsafe fn yield0([*mut (); 0]) -> [*mut (); 4];
+//   unsafe fn yield1([*mut (); 1]) -> [*mut (); 4];
+//   unsafe fn yield2([*mut (); 2]) -> [*mut (); 4];
+//   unsafe fn yield3([*mut (); 3]) -> [*mut (); 4];
+//   unsafe fn yield4([*mut (); 4]) -> [*mut (); 4];
+//   unsafe fn syscall0<const CLASS: usize>([*mut (); 0]) -> [*mut (); 4];
+//   unsafe fn syscall1<const CLASS: usize>([*mut (); 1]) -> [*mut (); 4];
+//   unsafe fn syscall2<const CLASS: usize>([*mut (); 2]) -> [*mut (); 4];
+//   unsafe fn syscall3<const CLASS: usize>([*mut (); 3]) -> [*mut (); 4];
+//   unsafe fn syscall4<const CLASS: usize>([*mut (); 4]) -> [*mut (); 4];
+//
+// However, not all of these are used! If we remove the system calls that are
+// unused, we are left with the following:
+//
+//   unsafe fn yield1([*mut (); 1]) -> [*mut (); 4];
+//   unsafe fn yield2([*mut (); 2]) -> [*mut (); 4];
+//   unsafe fn syscall1<const CLASS: usize>([*mut (); 1]) -> [*mut (); 4];
+//   unsafe fn syscall2<const CLASS: usize>([*mut (); 2]) -> [*mut (); 4];
+//   unsafe fn syscall4<const CLASS: usize>([*mut (); 4]) -> [*mut (); 4];
+//
+// To avoid making the RawSyscalls implementation index into arrays, we replace
+// the arrays in the input with multiple arguments. For symmetry, we also
+// replace the output with a tuple of individual values. This gives:
```

Comment:
Why is indexing into arrays a problem? You can just write stuff like
```rs
unsafe fn yield1([arg1]: [*mut (); 1]) -> ...;
```
Yay pattern matching!

Reply:
D'oh, I didn't think of that! I'll leave it as-is for a bit longer so reviewers can comment on whether they prefer individual values or arrays here. If I don't get any other input, I'm leaning towards switching back to arrays.

Reply:
If we switch back to arrays everywhere, including the functions with only 1 input register, then we can simplify the explanation of the design of `RawSyscalls`. I think that has value in and of itself.

Reply:
I made the switch to arrays. I switched all arguments, including those with 1 input register.